### DISCUSSION TITLE
How can I use UseEmbeddedResourcesProject in my unit test project?

### DISCUSSION THREAD

**Discussion by VictorioBerra** [toddams/RazorLight#418] [Q&A] [answered]
I have `UseEmbeddedResourcesProject(typeof(Program))'` in my main project, but in my unit tests, it throws:

`Could not find a part of the path 'C:\SourceCode\MyApp\tests\MyApp.IntegrationTests\bin\Debug\net5.0\Templates\GenericEmail.cshtml'.`

It should not be looking in `MyApp.IntegrationTests`.

My registration method:

```csharp
/// <summary>
/// Sets up FluentEmail and RazorLight for email templates.
/// </summary>
/// <param name="services">The IServiceCollection.</param>
/// <param name="hostingEnvironment">The IHostEnvironment.</param>
/// <param name="sendGridOptions">The SendGridOptions.</param>
/// <param name="embeddedResRootType">The root type for the project that has the embedded templates..</param>
/// <returns>The same IServiceCollection.</returns>
public static IServiceCollection AddEmail(
    this IServiceCollection services,
    IHostEnvironment hostingEnvironment,
    SendGridOptions sendGridOptions,
    Type embeddedResRootType) =>
    services.AddFluentEmail(sendGridOptions.EmailSender)
            .AddRazorRenderer(embeddedResRootType)
        .Services
            .AddIfElse(
                hostingEnvironment.IsDevelopment(),
                devServices => devServices.AddSingleton<ISender>(x => new SmtpSender(new SmtpClient("localhost", 25))),
                prodServices => prodServices.AddSingleton<ISender>(x => new SendGridSender(sendGridOptions.APIKey)));
```

My test looks similar to this:

```csharp
[Fact]
public void Should_Send_Email_and_Create_Output()
{
    var actualEmailSendTo = string.Empty;
    var actualEmailSendFrom = string.Empty;
    var actualSubject = string.Empty;

    // Mock the sender
    var mockEmailSender = new Mock<ISender>();
    mockEmailSender
        .Setup(x => x.SendAsync(It.IsAny<IFluentEmail>(), null))
        .Callback((IFluentEmail email, CancellationToken? token) =>
        {
            actualEmailSendTo = email.Data.ToAddresses.Single().EmailAddress;
            actualEmailSendFrom = email.Data.FromAddress.EmailAddress;
            actualSubject = email.Data.Subject;
        });

    var factory = new CustomWebApplicationFactory<Program>(this.testOutputHelper)
        .WithWebHostBuilder(builder =>
        {
            builder.ConfigureTestServices(services =>
            {
                services.AddSingleton(mockEmailSender.Object);
            });
        });

    // Start the server
    factory.CreateClient();
```

I show the above to show that CustomWebAppFactory<Program> is what launches the host and all that. So maybe thats part of the issue?

**Comment by jzabroski** [comment]
Did you look at our tests?

There is also the new Debug options added in rc3 as part of the RazorLightEngineBuilder.  Please try enabling that first.

**Reply by VictorioBerra** [reply]
Looking into it, looks like FluentEmail.Razor uses an older version of RazorLight where those debug options are not yet available.

https://github.com/lukencode/FluentEmail/blob/master/src/Renderers/FluentEmail.Razor/RazorRenderer.cs

I have created a PR in FluentEmail to allow me to provide the engine directly so I can get more control over the configuration. https://github.com/lukencode/FluentEmail/pull/240

**Comment by VictorioBerra** [comment, marked as answer]
Turns out in Fluent.Email, I was not using `.UsingTemplate` when I injected IFluentEmail into my service. So internally, it was using something other than my razor engine.

**Reply by VictorioBerra** [reply]
Although, I am having some issues with reference assemblies now.

**Reply by VictorioBerra** [reply]
My TEST project needed the following:

```
  <PropertyGroup Label="RazorLight">
    <!-- https://github.com/toddams/RazorLight#im-getting-cannot-find-reference-assembly-microsoftaspnetcoreantiforgerydll-exception-on-net-core-app-30-or-higher -->
    <PreserveCompilationReferences>true</PreserveCompilationReferences>
  </PropertyGroup>
  ```
  
  As seen on the README. The problem is I read the README, but without knowing why im adding these XML directives, its hard to connect the dots. Some parts of the readme explain the "whys" but not all of them.
